File: lesson_2/zad.cpp
#include<iostream>
#include<string>
#include<fstream>
#include<vector>

#define pb push_back

using namespace std;


struct arr{
    string name, surename;
    int year;
    float accuracy;
};


bool check_candidate(string name, string surename, int year, float accuracy){
    int name_size = name.size();
    int sname_size = surename.size();
    return (name_size > 3 && sname_size >= 3 && surename.substr( sname_size - 3, 3 ) == "ski");
}


string anonymization(string x){
    int n = x.size()-4;
    string new_x="";
    new_x += x[0];
    while(n--)new_x += '*';
    new_x += x[x.size()-3];
    new_x += x[x.size()-2];
    new_x += x[x.size()-1];
    return new_x;
}


int main(){
    vector<arr>output;

    string filename;
    cin>>filename;
    ifstream in(filename + ".txt", ios::in);

    if(in.fail()){
        cout<<"File does not exist"<<endl;
        return 1;
    }


    arr curr_max; curr_max.accuracy=-1.0;


    string name, surename;
    int year;
    float accuracy;

    while(in>>name>>surename>>year>>accuracy){

        if ( check_candidate(name, surename, year, accuracy)  && accuracy > curr_max.accuracy){
            curr_max.name = anonymization(name);
            curr_max.surename = anonymization(surename);
            curr_max.year = year;
            curr_max.accuracy = accuracy;
            output.clear();
            output.pb(curr_max);
        }
        else if ( check_candidate(name, surename, year, accuracy)  && accuracy == curr_max.accuracy){
            curr_max.name = anonymization(name);
            curr_max.surename = anonymization(surename);
            curr_max.year = year;
            curr_max.accuracy = accuracy;
            output.pb(curr_max);
        }
    }
    in.close();

    ofstream out(filename + ".max", ios::out);

    if(output.empty()){
        out<<endl;
        return 0;
    }
    for(int i=0; i<output.size(); i++){
        out<<output[i].name<<" "<<output[i].surename<<" "<<output[i].year<<endl;
    }
    out.close();

    return 0;
}

File: lesson_2/zad.py
def read_data(filename):
    arr = []

    with open(filename+'.txt') as file:
        for line in file:
            name = line.split(sep=' ')[0]
            surename = line.split(sep=' ')[1]
            year = int(line.split(sep=' ')[2])
            accuracy = float(line.split(sep=' ')[3])
            arr.append((name, surename, year, accuracy))
    return arr       


def write_data(filename, arr):
    with open(filename+'.max', 'w') as file:
        for data in arr:
            file.write(f'{anonymization(data[0])} {anonymization(data[1])} {data[2]}\n')


def anonymization(text):
    new_text = text[0] + '*'*(len(text)-4) + text[-3:]
    return new_text


def check(candidate):
    if(len(candidate[0])>3 and len(candidate[1])>=3 and candidate[1][-3:] == 'ski'):
        return 1
    return 0


filename = input()
table = read_data(filename)

output__table = []
curr_leader = ('', '', '', -1)

for candidate in table:
    if (check(candidate) and candidate[3] > curr_leader[3]):
        curr_leader = candidate
        output__table = []
        output__table.append(curr_leader)
    elif (check(candidate) and candidate[3] == curr_leader[3]):
        curr_leader = candidate
        output__table.append(curr_leader)



write_data(filename, output__table)